### DISCUSSION TITLE
Handle exceptions in an App so they don't crash all Apps.

### DISCUSSION THREAD

**Discussion by ngelotte** [net-daemon/netdaemon#241] [Feature requests]
Currently if an app throws an uncaught exception it crashes the entire netDaemon service. If another app is in the middle of something then it goes down as well and cannot complete. This is a problem when you have apps that have some waiting involved. For example a PIR light that is waiting for a lack of motion. If a different app crashes after the light has been turned on while it is waiting for more motion the light will just stay on because the timeout has ended.

**Comment by helto4real** [comment]
I will look into this, if you can provide with some code it would help finding the problem. To reproduce. Also bugs should be issues so please report it as but and provide code if you can. Thanks.

**Reply by ngelotte** [reply]
I thought I could make a simple example. But it just crashes the current app and leaves the other apps running. But I ran into a case where I thought the crash was causing netdaemon to start back up and reconnect. Unfortunately I cannot access old netdaemon logs so I will have to wait and see if it happen again.

**Reply by helto4real** [reply]
Yea I will add some more tests app crashes. I hade some crash myself but it happend like 2 times in one year so hard to reproduce. New bugreport #243 is created. Thanks for reporting.

**Reply by helto4real** [reply]
> I thought I could make a simple example. But it just crashes the current app and leaves the other apps running. But I ran into a case where I thought the crash was causing netdaemon to start back up and reconnect. Unfortunately I cannot access old netdaemon logs so I will have to wait and see if it happen again.

Do you have any example code??? Made 10 tests and still not breaks :)

**Reply by ngelotte** [reply]
I may have been wrong about what was happening as I might just have 2 colliding bugs. I have been having problems with my PIR sequence leaving the lights on. As I saw a bunch of errors from a different crash I thought the crash was causing the PIR turn off problem but maybe they are not related. Thanks for looking into this but unless I can find a more concrete example we can consider this closed. I am updating my apps to handle crashes internally because I need to run a shutdown sequence to make sure a crash does not leave some things on that should not be on. 

I am curious though- what is the strategy for an app crashing? Does netDaemon restart the app immediately? Does it keep track of how many times an app has crashed?

**Reply by helto4real** [reply]
The intended behavior is that apps will recover without restart. If an exception occurs it should log and keep subscribing for event changes. It is a good practice to handle errors your self though as with all programming. Just file a bug with log if something weird happens. Thanks for reporting though. I am planning to include more telemetry into ND to improve logging and error management. Ideas are welcome
